x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlConfigMigrator.java

```diff
@@ -257,6 +272,60 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
 
+    /**
+     * Update the persistent task parameters of datafeed and job tasks
+     * that are missing the fields added in v6.6. If a task exists with
+     * the missing field it must have been created in an earlier version
+     * and survived an elasticsearch upgrade.
+     *
+     * The return value is the same as the {@code currentTasks} parameter.
+     * Rather than a deep copy being made the tasks are modifed in place.
+     *
+     * @param jobs          Job configs
+     * @param datafeeds     Datafeed configs
+     * @param currentTasks  The persistent tasks
+     * @param nodes         The nodes in the cluster
+     * @return  The argument {@code currentTasks}
+     */
+    public static PersistentTasksCustomMetaData rewritePersistentTaskParams(Map<String, Job> jobs, Map<String, DatafeedConfig> datafeeds,
+                                                                            PersistentTasksCustomMetaData currentTasks,
+                                                                            DiscoveryNodes nodes) {
+
+        Collection<PersistentTasksCustomMetaData.PersistentTask> unallocatedJobTasks = MlTasks.unallocatedJobTasks(currentTasks, nodes);
+
+        for (PersistentTasksCustomMetaData.PersistentTask jobTask : unallocatedJobTasks) {
+            OpenJobAction.JobParams params = (OpenJobAction.JobParams) jobTask.getParams();
+            if (params.getJob() == null) {
+                Job job = jobs.get(params.getJobId());
+                if (job != null) {
+                    logger.debug("updating persistent task params for job [{}]", params.getJobId());
+                    params.setJob(job);
```

Comment:
Are you sure it's acceptable to update the parameters on `currentTasks`?  I'm not convinced that it is, because it's changing the _current_ local cluster state.  It means this node's opinion of the cluster state version `N` will be different from all the other nodes.  And version `N+1` may not even get broadcast to other nodes if the local node thinks it's identical to version `N`.

Unless I'm mistaken, I think you need to build a new `PersistentTasksCustomMetaData` object rather than updating the one from current state.

Reply:
I had the same reservations hence all the javadoc comments. There is an assertion in the full cluster restart test that the task params are actually updated so it does appear to work.

https://github.com/elastic/elasticsearch/blob/a3d0c2c91ff56ec4d246e853d7454f6111cdca2f/x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart/MlMigrationFullClusterRestartIT.java#L236

However reviewing `PersistentTasksCustomMetaData` uses the default implementation of `AbstractNamedDiffable.readDiffFrom` which returns the entire object and the TODO here https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/persistent/PersistentTasksCustomMetaData.java#L66 suggests that will not always be the case. Good catch I can't remember why I didn't use the builder in the first place